### ISSUE TITLE
"mod+Shift Space" binding also triggers "mod Space"

### ISSUE THREAD

**Issue by dnkl** [riverwm/river#604]
With this `init`:

```sh
#!/bin/bash

mod="Mod4"
riverctl map normal $mod Space spawn fuzzel
riverctl map normal $mod+Shift Space toggle-float

riverctl default-layout rivertile
exec rivertile
```
pressing `Mod4+Shift+Space` **both** toggles float, **and** spawns fuzzel.

River: [0.2.0-dev-90b3764](https://github.com/riverwm/river/commit/90b37645cf7744d8cc2a000fd3d74caf95bce204)

**Comment by dnkl** [comment]
XKB settings:
```sh
export XKB_DEFAULT_LAYOUT=se,us
export XKB_DEFAULT_OPTIONS=ctrl:swapcaps_and_switch_layout
```

**Comment by ifreund** [comment]
This is caused by #577.

Before that patch we would check keybinds first with modifiers taken into account. Then, if no match was found, we would check keybindings against the base layer as if modifiers had no effect. This allows the user to map `Mod4+Shift 1` instead of `Mod4+Shift !` on a US layout for example.

In #577 this behavior was changed to always check both and run all matching mappings even if the check with modifiers taken into account was successful.

I think the old behavior is more intuitive in this case, cc @pmkap.

**Comment by pmkap** [comment]
Apparently, I didn't take all things into account with my patch. I'm going to look into it.

**Comment by pmkap** [comment]
Apparently, some layouts behave differently when pressing `Shift Space`. It seems that [`xkb_state_key_get_consumed_mods2()`](https://xkbcommon.org/doc/current/group__state.html#ga5b119b99542fcad1787119191ba9e286) reports `Shift` as consumed in this case, so that `Shift Space` translates to `Space`. Not only on the Swedish layout but also others. Finnish, Norwegian and Indian for example.

Before #577, we didn't have this problem because we didn't translate (consume) the modifiers. This was a `TODO` https://github.com/riverwm/river/blob/ae349b0ce45b44b79a0a93e89c2cd05d24bb63d4/river/Keyboard.zig#L88 

Sway translates the modifiers, so `Mod4+Shift Space` triggers a mapping for `Mod4 Space` on the above mentioned layouts. Only that sway never executes two mappings so probably nobody ever noticed this. 

So the way I see it we can either not translate modifiers like it was before #577. Or we need to check mappings without translating the keys *first*.

**Comment by ifreund** [comment]
@pmkap Thanks for that investigation, looks like my understanding in the above comment was incomplete, who knew that `Shift Space` had a different meaning in Swedish :D

Is there a downside to going back to the pre #577 behavior and not translating the modifiers? That seems like the most predictable option with this new knowledge  of how strangely xkb layouts can behave.

**Comment by pmkap** [comment]
> Is there a downside to going back to the pre #577 behavior and not translating the modifiers? That seems like the most predictable option with this new knowledge of how strangely xkb layouts can behave.

The downside is that when you create a mapping e.g. `map normal Super Exclam spawn foot`. This mapping *cannot* be triggered (on US layout) before #577. I.e. when pressing `Super+Shift 1` (physical keys) you would expect this mapping to be triggered because `Shift` gets "consumed" in the process of translating `1` to `Exclam` (Current/Sway behavior). However, pre #577, we checked the following:

1. Translate `Shift+1` to `Exclam` but **without** consuming `Shift`. So we checked for a mapping `Super+Shift Exclam`.
2. Don't translate anything, just check for a `Super+Shift 1` mapping.

Although this is not really "correct", is prevents the awkward behavior reported by dnkl.

**Comment by pmkap** [comment]
And it worked before, nobody ever complained, right?. The change is simple, I could prepare a patch. Maybe with a comment explaining why we are not "consuming" modifiers.

**Comment by ifreund** [comment]
> 1. Translate Shift+1 to Exclam but without consuming Shift. So we checked for a mapping Super+Shift Exclam.
> 2. Don't translate anything, just check for a Super+Shift 1 mapping.

Hmm, this makes me wonder if we should even bother checking 1. It seems to me that allowing mapping of any keysym on the base xkbcommon level plus any combination of modifiers would be sufficient.

In fact, I question whether it would have even been possible for the first check we were doing to match anything.

> And it worked before, nobody ever complained, right?

Indeed, I think we should go back to the previous behavior and stop translating keysyms for the purpose of mappings. Thanks for the detailed explanation!

**Comment by pmkap** [comment]
> In fact, I question whether it would have even been possible for the first check we were doing to match anything.

It is was possible, e.g. in the above example (pressing `Super+Shift 1`), a `Super+Shift Exclam` mapping would have been executed, even before/instead of  a `Super+Shift 1` mapping. But I agree that this "incomplete" translation is not intuitive and we might as well skip the translation completely.

**Comment by ifreund** [comment]
This should be fixed by 47c02ebcbc484e44db728a84e446e3a159474e71.

Thanks for the bug report @dnkl, without your swedish layout it would have taken us much longer to discover this!
